main.cpp

```diff
@@ -180,139 +123,268 @@ int main()
+
+    FloatType(float v) : value(v) {}
 };
 
-float FloatType::add(float lhs, float rhs)
+FloatType& FloatType::add(float rhs)
 {
-    return lhs + rhs;
+    value += rhs;
+    return *this;
 }
 
-float FloatType::subtract(float lhs, float rhs)
+FloatType& FloatType::subtract(float rhs)
 {
-    return lhs - rhs;
+    value -= rhs;
+    return *this;
 }
 
-float FloatType::multiply(float lhs, float rhs)
+FloatType& FloatType::multiply(float rhs)
 {
-    return lhs * rhs;
+    value *= rhs;
+    return *this;
 }
 
-float FloatType::divide(float lhs, float rhs)
+FloatType& FloatType::divide(float rhs)
 {
     // warn the user if they're doing floating-point-division-by-zero 
     // but not prevent the division from happening
     if (rhs == 0.f) 
-        std::cout << "\nwarning, floating point division by zero returns 'inf' !\n";        
-    return lhs / rhs;
+        std::cout << "warning: floating point division by zero!\n";        
+    value /= rhs;
+    return *this;
 }
 
 struct DoubleType
 {
-    double add(double lhs, double rhs);
-    double subtract(double lhs,double rhs );
-    double multiply(double lhs, double rhs );
-    double divide(double lhs,double rhs );
+    double value;
+
+    DoubleType& add(double rhs);
+    DoubleType& subtract(double rhs );
+    DoubleType& multiply(double rhs );
+    DoubleType& divide(double rhs );
+
+    DoubleType& add(const DoubleType& dt);
+    DoubleType& subtract(const DoubleType& dt);
+    DoubleType& multiply(const DoubleType& dt);
+    DoubleType& divide(const DoubleType& dt);
+
+    DoubleType& add(const IntType& dt);
+    DoubleType& subtract(const IntType& dt);
+    DoubleType& multiply(const IntType& dt);
+    DoubleType& divide(const IntType& dt);
+
+    DoubleType& add(const FloatType& dt);
+    DoubleType& subtract(const FloatType& dt);
+    DoubleType& multiply(const FloatType& dt);
+    DoubleType& divide(const FloatType& dt);
+
+    DoubleType(double d) : value(d) {}
 };
 
-double DoubleType::add(double lhs, double rhs)
+DoubleType& DoubleType::add(double rhs)
 {
-    return lhs + rhs;
+    value += rhs;
+    return *this;
 }
 
-double DoubleType::subtract(double lhs, double rhs)
+DoubleType& DoubleType::subtract(double rhs)
 {
-    return lhs - rhs;
+    value -= rhs;
+    return *this;
 }
 
-double DoubleType::multiply(double lhs, double rhs)
+DoubleType& DoubleType::multiply(double rhs)
 {
-    return lhs * rhs;
+    value *= rhs;
+    return *this;
 }
 
-double DoubleType::divide(double lhs, double rhs)
+DoubleType& DoubleType::divide(double rhs)
 {
     if (rhs == 0.) 
-        std::cout << "\nwarning, floating point division by zero returns 'inf' !\n";
-    return lhs / rhs;
+        std::cout << "warning: floating point division by zero!\n";
+    value /= rhs;
+    return *this;
 }
 
 struct IntType
 {
-    int add(int lhs, int rhs);
-    int subtract(int lhs,int rhs );
-    int multiply(int lhs, int rhs );
-    int divide(int lhs,int rhs );
+    int value;
```

Comment:
This does not demonstrate the following:
```
Edit your 3 structs so that they own a heap-allocated primitive type 
without using smart pointers named 'value'
```
- [ ] revise